internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+                    f,
+                    "forall [{}] where {} {}",
+                    List { values: &self.free },
+                    List { values: &bounds },
+                    self.expr,
+                )
+            }
+            None => write!(f, "forall [{}] {}", List { values: &self.free }, self.expr),
+        }
+    }
+}
+
+// Kind represents a class or family of types
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Kind {
+    Add,
+    Div,
+    Cmp,
+    Nul,
+}
+
+impl fmt::Display for Kind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Kind::Add => f.write_str("Addable"),
+            Kind::Div => f.write_str("Divisible"),
+            Kind::Cmp => f.write_str("Comparable"),
+            Kind::Nul => f.write_str("Nullable"),
+        }
+    }
+}
+
+// MonoType represents a specific named type
+#[derive(Debug, Clone, PartialEq)]
+pub enum MonoType {
+    Bool,
+    Int,
+    Uint,
+    Flt,
+    Str,
+    Dur,
+    Time,
+    Rgx,
+    Var(Tvar),
+    Arr(Box<Array>),
+    Rec(Box<Row>),
+    Fun(Box<Function>),
+}
+
+impl fmt::Display for MonoType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MonoType::Bool => f.write_str("bool"),
+            MonoType::Int => f.write_str("int"),
+            MonoType::Uint => f.write_str("uint"),
+            MonoType::Flt => f.write_str("flt"),
+            MonoType::Str => f.write_str("str"),
+            MonoType::Dur => f.write_str("dur"),
+            MonoType::Time => f.write_str("time"),
+            MonoType::Rgx => f.write_str("rgx"),
+            MonoType::Var(var) => var.fmt(f),
+            MonoType::Arr(arr) => arr.fmt(f),
+            MonoType::Rec(obj) => obj.fmt(f),
+            MonoType::Fun(fun) => fun.fmt(f),
+        }
+    }
+}
+
+// Tvar stands for type variable.
+// A type variable holds an unknown type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Tvar(i64);
+
+impl fmt::Display for Tvar {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "t{}", self.0)
+    }
+}
+
+// Array is a homogeneous list type
+#[derive(Debug, Clone, PartialEq)]
+pub struct Array(MonoType);
+
+impl fmt::Display for Array {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[{}]", self.0)
+    }
+}
+
+// Row is an extensible record type.
+//
+// A row is either Nil meaning it has no properties,
+// or it is an extension of a row.
+//
+// A row may extend what is referred to as a row
+// variable. A row variable is a type variable that
+// represents an unknown record type.
+//
+#[derive(Debug, Clone)]
+pub enum Row {
+    Nil,
+    Var(Tvar),
+    Extension { head: Property, tail: Box<Row> },
+}
+
+impl fmt::Display for Row {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Row::Nil => f.write_str("{}"),
+            Row::Var(tv) => tv.fmt(f),
```

Comment:
Since this is a row variable should it be wrapped in `{}` for formatting?

Reply:
That would be unconventional so I would prefer not to.

Reply:
Sounds good